src/NServiceBus.Transport.Msmq/Persistence/SubscriptionStorage/MsmqSubscriptionStorage.cs

```diff
@@ -62,7 +64,7 @@ public void Dispose()
         public Task<IEnumerable<Subscriber>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes, ContextBag context, CancellationToken cancellationToken = default)
         {
             var messagelist = messageTypes.ToArray();
-            var result = new HashSet<Subscriber>();
+            var result = new HashSet<Subscriber>(SubscriberComparer);
```

Comment:
While this code could have returned duplicate subscribers, there's code in NServiceBus's UnicastPublishRouter class in SelectDestinationsForEachEndpoint() that dedupes destination addresses, making this more of a refactoring, and not a bug that needs to be patched.